FeedStoreChallenge/CoreData/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,125 @@
+import Foundation
+import CoreData
+
+public class CoreDataFeedStore: FeedStore {
+
+    private let context: NSManagedObjectContext
+
+    public init(storeURL: URL) throws {
+        let container = try PersistentContainer(storeURL: storeURL)
+        self.context = container.newBackgroundContext()
+    }
+
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+
+        let context = self.context
+        context.perform { [weak self] in
+
+            guard let fetchedFeed = self?.fetchFeed() else {
+                return completion(.empty)
+            }
+
+            let images = fetchedFeed.images?.array as? [PersistentFeedImage] ?? []
+            let timestamp = fetchedFeed.timestamp ?? Date()
+
+            completion(.found(feed: images.toLocalFeedImage(), timestamp: timestamp))
+        }
+    }
+
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+
+        let context = self.context
+        context.perform { [weak self] in
+
+            guard let fetchedFeed = self?.fetchFeed() else {
+                return completion(nil)
+            }
+            self?.deleteFeed(fetchedFeed)
```

Comment:
The `deleteFeed` method doesn't make the code any more clear than using `context.delete(feed)`.

It could be just `context.delete(feed)` so you also avoid capturing `self` within this closure.

```suggestion
            context.delete(feed)
```